Update README (#43)
FreeRTOS/coreMQTT-Agent-Demos: README.md

## MQTT Agent and Demo Tasks (including OTA) - Using coreMQTT

> [coreMQTT](https://github.com/FreeRTOS/coreMQTT) is an MIT licensed open source C MQTT client library for microcontrollers and small microprocessor based IoT devices. Its design is intentionally simple to ensure it has no dependency on any other library or operating system, and to better enable static analysis including [memory safety proofs](https://www.freertos.org/2020/02/ensuring-the-memory-safety-of-freertos-part-1.html). That simplicity and lack of operating system dependency (coreMQTT does not require multithreading at all) means coreMQTT does not build thread safety directly into its implementation. Instead thread safety must be provided by higher level software. This labs project implements a coreMQTT extension that provides that higher level functionality in the form of an MQTT agent (or MQTT daemon). While the implementation demonstrated here is currently specific to FreeRTOS, there are not many dependencies on FreeRTOS, meaning the implementation can easily be adapted for use with other operating systems.

Reviewer: Is this text relevant for this repository? Shouldn't it just say that this repository provides demo examples of using the coreMQTT Agent library to share a single MQTT connection amongst multiple application tasks? 
Maybe, after that a brief paragraph about the coreMQTT Agent library can be provided.

Author: Added a few sentences before about this repository, but left this block here in a subsection

> ## MQTT Agent and Demo Tasks (including OTA) - Using coreMQTT

Reviewer 2: coreMQTT Agent?

Author: Reworded

> This repository provides demonstration examples of using the [coreMQTT Agent library](https://github.com/FreeRTOS/coreMQTT-Agent) with FreeRTOS. The coreMQTT Agent library is an extension on top of [coreMQTT](https://github.com/FreeRTOS/coreMQTT) that provides MQTT APIs with thread safety. The examples here share a single MQTT connection amongst multiple concurrent tasks, without requiring the management of any synchronization primitives from the application.

Reviewer 2: I'd start with the most important points about the demo. For ex: "the demo uses OTA, Device Shadow, and Device Defender in multiple threads."

Author: Reworded

> ### What is coreMQTT?

Reviewer 2: Why is this question relevant here? A user trying to use the demo would like to see info on coreMQTT-Agent rather than coreMQTT?

Author: I will remove this paragraph as I think the above section has sufficient information on coreMQTT Agent's thread safety

## MQTT Agent and Demo Tasks (including OTA) - Using coreMQTT

This repository provides demonstration examples of using the [coreMQTT Agent library](https://github.com/FreeRTOS/coreMQTT-Agent) with FreeRTOS. The coreMQTT Agent library is an extension on top of [coreMQTT](https://github.com/FreeRTOS/coreMQTT) that provides MQTT APIs with thread safety. The examples here share a single MQTT connection amongst multiple concurrent tasks, without requiring the management of any synchronization primitives from the application.

### What is coreMQTT?

> [coreMQTT](https://github.com/FreeRTOS/coreMQTT) is an MIT licensed open source C MQTT client library for microcontrollers and small microprocessor based IoT devices. Its design is intentionally simple to ensure it has no dependency on any other library or operating system, and to better enable static analysis including [memory safety proofs](https://www.freertos.org/2020/02/ensuring-the-memory-safety-of-freertos-part-1.html). That simplicity and lack of operating system dependency (coreMQTT does not require multithreading at all) means coreMQTT does not build thread safety directly into its implementation. Instead thread safety must be provided by higher level software. This project implements a coreMQTT extension that provides that higher level functionality in the form of an MQTT agent (or MQTT daemon). While the implementation demonstrated here is currently specific to FreeRTOS, there are not many dependencies on FreeRTOS, meaning the implementation can easily be adapted for use with other operating systems.

Reviewer: The last 2 sentences in this paragraph are not relevant to coreMQTT. 
```suggestion
[coreMQTT](https://github.com/FreeRTOS/coreMQTT) is an MIT licensed open source C MQTT client library for microcontrollers and small microprocessor based IoT devices. Its design is intentionally simple to ensure it has no dependency on any other library or operating system, and to better enable static analysis including [memory safety proofs](https://www.freertos.org/2020/02/ensuring-the-memory-safety-of-freertos-part-1.html). That simplicity and lack of operating system dependency (coreMQTT does not require multithreading at all) means coreMQTT does not build thread safety directly into its implementation. Instead thread safety must be provided by higher level software. The coreMQTT Agent library represents an extension to the coreMQTT library to provide the higher level thread-safe MQTT API functionality in the form of an MQTT agent (or MQTT daemon). ```

Author: I'm going to edit your comment for `represents` -> `is` and since you needed a newline before "```"

To clone using HTTPS:
```
git clone https://github.com/FreeRTOS/coreMQTT-Agent-Demos.git --recurse-submodules
```
Using SSH:
```
git clone git@github.com:FreeRTOS/coreMQTT-Agent-Demos.git --recurse-submodules
```

If you have downloaded the repo without using the `--recurse-submodules` argument, you need to run:
```
git submodule update --init --recursive
```

## Getting started

> The [documentation page](https://freertos.org/mqtt/mqtt-agent-demo.html) for this repository contains information on the MQTT agent and the contained demo project.  There is also a [supplemental documentation page](https://freertos.org/ota/ota-mqtt-agent-demo.html) that describes how to run an Over the Air (OTA) update agent as one of the RTOS tasks that share the same MQTT connection.

Reviewer: OTA goes as **Over-the-Air** in the branding: https://docs.aws.amazon.com/freertos/latest/userguide/freertos-ota-dev.html
```suggestion
The [documentation page](https://freertos.org/mqtt/mqtt-agent-demo.html) for this repository contains information on the MQTT agent and the contained demo project.  There is also a [supplemental documentation page](https://freertos.org/ota/ota-mqtt-agent-demo.html) that describes how to run an Over-the-Air (OTA) update agent as one of the RTOS tasks that share the same MQTT connection.
```

## Building Demos
This repository contains build files for the FreeRTOS Windows port - able to be run with the free [Community version of Visual Studio](https://visualstudio.microsoft.com/vs/community/) - and an ARM Cortex M model in the [QEMU hardware emulator](https://www.qemu.org/).

### Visual Studio

1. From the Visual Studio IDE, open the `mqtt_multitask_demo.sln` Visual Studio solution file in the `build/VisualStudio/` directory.

1. Select **Build Solution** from the IDE's **Build** menu

> ### QEMU Cortex-M3

Reviewer: Will there be no instructions on how to run the executable in QEMU? It could be helpful for the reader to be pointed to a link for using QEMU

Author: I would add it if I knew, but unfortunately I have not tried running it

Reviewer 3: Added.

**Command Line**

1. Ensure that `arm-none-eabi-gcc` from the [GNU ARM Embedded Toolchain](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads) and GNU `make` are in your path.

1. From the `build/Cortex-M3_MPS2_QEMU_GCC` directory, run `make`.

**Eclipse IDE**

1. Ensure that `arm-none-eabi-gcc` from the [GNU ARM Embedded Toolchain](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads) and GNU `make` are in your path.

1. Open the `build/Cortex-M3_MPS2_QEMU_GCC` directory as an Eclipse project.

1. Select `Project -> Build Project` from the Eclipse menu.

> In both cases, the generated executable can be found at `build/Cortex-M3_MPS2_QEMU_GCC/output/RTOSDemo.elf`.

Reviewer: ```suggestion
In both cases of using **QEMU Cortex-M3** or **Eclipse IDE**, the generated executable can be found at `build/Cortex-M3_MPS2_QEMU_GCC/output/RTOSDemo.elf`.
```

To clone using HTTPS:
```
git clone https://github.com/FreeRTOS/coreMQTT-Agent-Demos.git --recurse-submodules
```
Using SSH:
```
git clone git@github.com:FreeRTOS/coreMQTT-Agent-Demos.git --recurse-submodules
```

If you have downloaded the repo without using the `--recurse-submodules` argument, you need to run:
```
git submodule update --init --recursive
```

## Getting started

> The [documentation page](https://freertos.org/mqtt/mqtt-agent-demo.html) for this repository contains information on the MQTT agent and the contained demo project.  There is also a [supplemental documentation page](https://freertos.org/ota/ota-mqtt-agent-demo.html) that describes how to run an Over-the-Air (OTA) update agent as one of the RTOS tasks that share the same MQTT connection.

Author: I believe this is the new URL
```suggestion
The [documentation page](https://freertos.org/mqtt-agent/mqtt-agent-demo.html) for this repository contains information on the MQTT agent and the contained demo project.  There is also a [supplemental documentation page](https://freertos.org/ota/ota-mqtt-agent-demo.html) that describes how to run an Over-the-Air (OTA) update agent as one of the RTOS tasks that share the same MQTT connection.
```

## Building Demos

> This repository contains both a Visual Studio project that uses the FreeRTOS Windows port and can be built using the free [Community version of Visual Studio](https://visualstudio.microsoft.com/vs/community/), and a GCC/makefile project that build the FreeRTOS ARM Cortex-M3 port and targets the [QEMU hardware emulator](https://www.qemu.org/).  The makefile can be built form the command line or the provided Eclipse project, and works on both Windows and Linux hosts.

Author: ```suggestion
This repository contains both a Visual Studio project that uses the FreeRTOS Windows port and can be built using the free [Community version of Visual Studio](https://visualstudio.microsoft.com/vs/community/), and a GCC/makefile project that build the FreeRTOS ARM Cortex-M3 port and targets the [QEMU hardware emulator](https://www.qemu.org/).  The makefile can be built from the command line or the provided Eclipse project, and works on both Windows and Linux hosts.
```

### Visual Studio

1. Ensure to follow the instructions on the above linked demo documentation page to configure the build as required to access your network.

1. From the Visual Studio IDE, open the `mqtt_multitask_demo.sln` Visual Studio solution file in the `build/VisualStudio/` directory.

1. Select **Build Solution** from the IDE's **Build** menu.

### QEMU Cortex-M3

**Command Line**

1. Ensure that `arm-none-eabi-gcc` from the [GNU ARM Embedded Toolchain](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads) and GNU `make` are in your path.

1. From the `build/Cortex-M3_MPS2_QEMU_GCC` directory, run `make`.

**Eclipse IDE**

1. Ensure that `arm-none-eabi-gcc` from the [GNU ARM Embedded Toolchain](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads) and GNU `make` are in your path.

1. Open the `build/Cortex-M3_MPS2_QEMU_GCC` directory as an Eclipse project.

1. Select `Project -> Build Project` from the Eclipse menu.

In both cases, the generated executable can be found at `build/Cortex-M3_MPS2_QEMU_GCC/output/RTOSDemo.elf`.

> **Running the makefile build in QEMU**

Author: Might as well capitalize this here since this looks like a section title, and we also capitalize the Makefile file in the directory
```suggestion
**Running the Makefile build in QEMU**
```

> Use the QEMU command contained in the [start_qemu.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu.bat) batch file to run the generated `RTOSDemo.elf` file directly.  Alternatively, use the command contained in the [start_qemu_and_wait.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu_and_wait.bat) batch file to start QEMU and wait for a GBD connection on port 1234.  The Eclipse project contains a suitable debug configuration.

Author: ```suggestion
Use the QEMU command contained in the [start_qemu.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu.bat) batch file to run the generated `RTOSDemo.elf` file directly.  Alternatively, use the command contained in the [start_qemu_and_wait.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu_and_wait.bat) batch file to start QEMU and wait for a GDB connection on port 1234.  The Eclipse project contains a configuration suitable for debugging.
```

### Visual Studio

1. From the Visual Studio IDE, open the `mqtt_multitask_demo.sln` Visual Studio solution file in the `build/VisualStudio/` directory.

1. Select **Build Solution** from the IDE's **Build** menu

### QEMU Cortex-M3

**Command Line**

1. Ensure that `arm-none-eabi-gcc` from the [GNU ARM Embedded Toolchain](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads) and GNU `make` are in your path.

1. From the `build/Cortex-M3_MPS2_QEMU_GCC` directory, run `make`.

**Eclipse IDE**

1. Ensure that `arm-none-eabi-gcc` from the [GNU ARM Embedded Toolchain](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads) and GNU `make` are in your path.

1. Open the `build/Cortex-M3_MPS2_QEMU_GCC` directory as an Eclipse project.

1. Select `Project -> Build Project` from the Eclipse menu.

In both cases, the generated executable can be found at `build/Cortex-M3_MPS2_QEMU_GCC/output/RTOSDemo.elf`.

## Getting help

> You can use your Github login to get support from both the FreeRTOS community and directly from the primary FreeRTOS developers on our [active support forum](https://forums.freertos.org).  The [FAQ](https://www.freertos.org/FAQ.html) provides another support resource.

Reviewer 2: "The FAQ provides another support resource" looks weird. Can we reword this to something like this? "You can find a list of frequently asked questions here."

Author: Reworded

In both cases, the generated executable can be found at `build/Cortex-M3_MPS2_QEMU_GCC/output/RTOSDemo.elf`.

**Running the makefile build in QEMU**

> Use the QEMU command line contained in the [start_qemu.bat](https://github.com/FreeRTOS/coreMQTT-Agent-Demos/blob/main/build/Cortex-M3_MPS2_QEMU_GCC/start_qemu.bat) batch file to run the generated 'RTOSDemo.elf' file directly.  Alternatively use the command line contained in the [start_qemu_and_wait.bat](https://github.com/FreeRTOS/coreMQTT-Agent-Demos/blob/main/build/Cortex-M3_MPS2_QEMU_GCC/start_qemu_and_wait.bat) batch file to start QEMU and wait for a GBD connection on port 1234.  The Eclipse project contains a suitable debug configuration.

Author: We should refer to it with a relative link since it's in the same repository. Also "QEMU command line" - can this just be "QEMU command"?
```suggestion
Use the QEMU command contained in the [start_qemu.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu.bat) batch file to run the generated `RTOSDemo.elf` file directly.  Alternatively, use the command contained in the [start_qemu_and_wait.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu_and_wait.bat) batch file to start QEMU and wait for a GBD connection on port 1234.  The Eclipse project contains a suitable debug configuration.
```

Author: Should this be GBD or GDB? Also, what is the Eclipse debug configuration suitable for? Should it instead be "configuration suitable for debugging"?
```suggestion
Use the QEMU command contained in the [start_qemu.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu.bat) batch file to run the generated `RTOSDemo.elf` file directly.  Alternatively, use the command contained in the [start_qemu_and_wait.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu_and_wait.bat) batch file to start QEMU and wait for a GDB connection on port 1234.  The Eclipse project contains a configuration suitable for debugging.
```

> Note these QEMU command lines assume there is a network tap driver called "TAP0".

Author: Could they just be referred to as commands?
```suggestion
Note that these QEMU commands assume there is a network tap driver called "TAP0".
```

> **Obtaing network access from QEMU on windows**

Author: ```suggestion
**Obtaining network access from QEMU on Windows**
```

1. Install the [OpenVPN TAP driver for Windows](https://openvpn.net/).

> 1. Name the installed TAP interface "TAP0".  This can be done by selecting "Change adaptor settings" from the Windows' Network Properties window (Control Panel\Network and Internet\Network Connections) to view all network adaptors, selecting the installed TAP adapter, then pressing F2 to edit the adaptor's name.

Author: There are two spellings of adapter here. I believe OpenVPN and Windows both use "e"
```suggestion
1. Name the installed TAP interface "TAP0".  This can be done by selecting "Change adapter settings" from Windows' Network Properties window (Control Panel\Network and Internet\Network Connections) to view all network adapters, selecting the installed TAP adapter, then pressing F2 to edit the adapter's name.
```

> 1. From the same window, configure a real wired network interface to have a static IP address that is compatible with whichever IP address would otherwise be allocated by a DHCP server on your local network.

Author: This is a little confusing. Does this mean get an address from the DHCP server and assign it as a static address, choose an available IP address from the DHCP server's range, or choose an IP address that does **not** conflict with the DHCP server's range? It sounds like it could be the second one, but I'm not sure
```suggestion
1. From the same window, configure a real wired network interface to have a static IP address that is within the range of IP addresses allocated by your local DHCP server.
```

> 1. Still in the same window, select both the real network adaptor and the TAP adapter at the same, then right click, and select "Bridge Connections" from the pop up menu.

Author: Both adaptor and adapter here as well.
```suggestion
1. Still in the same window, select both the real network adapter and the TAP adapter at the same, then right click, and select "Bridge Connections" from the pop up menu.
```
